File: 155-min-stack/min-stack.cpp
class MinStack {
public:
    std::stack<pair<int,int>> st;
    int min_elem;
    MinStack() { 
        min_elem = INT_MAX;    
    }
    
    void push(int val) {
        if (st.empty()) min_elem = val;
        else min_elem = min(st.top().second, val);
        st.push({val, min_elem});      
    }
    
    void pop() {
         st.pop();   
    }

    int top() {
        return st.top().first;      
    }
    
    int getMin() {
        return st.top().second;
        
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 155-min-stack/min-stack.py
class MinStack:

    def __init__(self):
        self.st = []
        self.min_elem = float('inf')      

    def push(self, val: int) -> None:
        if len(self.st)==0: self.min_elem = val
        else: self.min_elem = min(self.st[-1][1], val)
        self.st.append([val,self.min_elem])
        

    def pop(self) -> None:
        self.st.pop()

    def top(self) -> int:
        return self.st[-1][0]
        

    def getMin(self) -> int:
        return self.st[-1][1]

        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()